db.go

```diff
@@ -791,8 +771,357 @@ var eventDefs = map[EventType]EventDef{
 			return b, nil
 		},
 	},
+	// Send and Receive RPCs share table to make
+	EventTypeSendRPC: {
+		Name: "rpc_event",
+		DDL:  CreateRPCsTable,
+		BatchInsert: func(ctx context.Context, evs []*TraceEvent) (*pgx.Batch, error) {
+			logger := slog.With("event_type", "sent_rpc")
+			b := new(pgx.Batch)
+
+			// get the type of RPC that we are have
+			rpcDirection := SentRPC
+			parentCols := []string{"peer_id", "timestamp", "remote_peer_id", "direction"}
+
+			eventCount := 0
+			for _, ev := range evs {
+				if ev.Timestamp == nil {
+					logger.Debug("skipping event, no timestamp")
+					continue
+				}
+				// ID of the peer submitting traces
+				peerID, err := peer.IDFromBytes([]byte(ev.PeerID))
+				if err != nil {
+					logger.Debug("skipping event, bad peer id", err, "peer_id", ev.PeerID)
+					continue
+				}
+				// Get Parent info
+				remotePeerId, err := peer.IDFromBytes([]byte(ev.SendRPC.SendTo))
+				if err != nil {
+					logger.Debug("skipping event, bad peer id", err, "peer_id", ev.PeerID)
+					continue
+				}
+				// Check the type of subevent that we have received
+				var mgs []*MessageMetaEvent
+				var ihaves []*ControlIHaveMetaEvent
+				var iwants []*ControlIWantMetaEvent
+				childLen := 0
+				childColsLen := 0
+
+				msgType := ev.SendRPC.Meta.Type()
+				switch msgType {
+				case BroadcastMessage:
+					mgs = ev.SendRPC.Meta.Messages
+					if len(mgs) > 0 {
+						childLen = len(mgs)
+						childColsLen = 3 // (rpc_event_id, topic, message_id)
+					}
+				case IhaveMessage:
+					ihaves = ev.SendRPC.Meta.Control.Ihave
+					if len(ihaves) > 0 {
+						ihLen := 0
+						for _, ih := range ihaves {
+							ihLen += len(ih.MessageIDs)
+						}
+						childLen = len(ihaves) + ihLen
+						childColsLen = 3 // (rpc_event_id, topic, message_id)
+					}
+				case IwantMessage:
+					iwants = ev.SendRPC.Meta.Control.Iwant
+					if len(iwants) > 0 {
+						iwLen := 0
+						for _, iw := range iwants {
+							iwLen += len(iw.MessageIDs)
+						}
+						childLen = len(iwants) + iwLen
+						childColsLen = 2 // (rpc_event_id, message_id)
+					}
+				case UntraceableMessage:
+					logger.Warn("received unsupported SendRPC event", ev)
+					continue
+				}
+				// append parents values
+				values := make([]any, 0, len(parentCols)+childLen*childColsLen)
+				values = append(values,
+					peerID.String(),
+					time.Unix(0, *ev.Timestamp),
```

Comment:
can ev.Timestamp be nil? This will panic if it is, Can we test for that case before dereferencing the pointer

Reply:
All the `pubsub` traces come timestamped ([link](https://github.com/libp2p/go-libp2p-pubsub/blob/master/trace.go#L41)), so "it shouldn't happen". 
Although, to prevent the edgy cases, should we check and drop traces that do not come timestamped? All the `BatchInsert` functions assume that a valid pointer defines the timestamp.